### ISSUE TITLE
Merge C338 compatible library

### ISSUE THREAD

**Issue by joopert** [joopert/nad_receiver#13]
Copied from https://github.com/joopert/nad_receiver/issues/12#issuecomment-404127616
> I've made a C338 compatible library based on this repo.
It's available on gitlab: https://gitlab.com/mindig.marton/nadtcp

> It'd be great if it could be merged, so we don't have to create a third NAD component in home-assistant.
Though the API uses asyncio and no polling, so the interfaces need some tweaking to support both polling and non-polling protocols, so the home assistant component can handle both (as the D7050 cannot handle persistent connections)

@mindigmarton I created a new issue for this.
I agree that it would be great if it could be merged. Are you willing to create a pull request for this?

**Comment by mindigmarton** [comment]
Yeah, sure! I'll have to convert the existing TCP class to be async (but not asyncio).
Reconnection also needs to be handled here, so the HASS component doesn't have to know about it (my current implementation handles reconnections in HASS: [nadtcp2.py](https://gitlab.com/mindig.marton/ha-nadtcp/blob/master/components/media_player/nadtcp2.py) )

I could either set a timer in the async loop, or just reconnect when hass calls update, not sure which would be better.

**Comment by mindigmarton** [comment]
A little update: I had to refactor the C338 protocol class to provide a unified interface. There are some quirks:
- Auto standby is prevented if you poll the status of the device. Also, polling sends status updates to _everyone_, polling is pretty much broken for the C338.
- Disconnection is not detected until the socket is reset if you don't poll the device. (this is pretty normal, but in HASS, the device won't become unavailable)
- State updates have to be batched to not spam updates and this has to be handled client-side
- Responses are not reliable. Sometimes the C338 forgets to respond to 'Main?' query, for example, when you're adjusting volume with the physical knob simultaneously. The C338 API is fully async anyway, so there is no point in waiting for responses.

I refactored the D7050 to use asyncio, I've tested it with a C338 stub, but I can't test it with a D7050. It's pushed to my fork (the C338 version has changed quite a bit since):
https://github.com/mindigmarton/nad_receiver/tree/c338-asyncio

It should work pretty much the same, I just changed _send to async. 
Also, the D7050 also has an auto standby feature. Do you know if polling prevents standby like the C338?

I've been running the new API with HASS for a few weeks. The only issue left, that I want to fix before opening a PR is detecting socket disconnection without polling, and the D7050 version has to be tested.

**Comment by gladhorn** [comment]
Hi, I guess the attempts at merging this have died? It would be nice to get the support in. I think there are a bunch of nice additions in the fork. Switching everything over to asyncio is of course completely incompatible with how things are at the moment. I also considered doing that, but for now refrained from it. I'd suggest submitting small bits as PRs, I had to also fix sources being returned as string, so that would have saved me some effort, had I known that the work was already done somewhere else.

**Comment by mindigmarton** [comment]
Hey.
I got stuck at writing a common API for the telnet / tcp transports, since I couldn't test the telnet ones.
Also I got fed up with the C338 TCP API because it has a number of quirks and reading the D7050 API it has different behavior, so I just kept on trying to improve the C338 integration first.

For example, the C338 sends broadcast messages, and there is no message correlation, so if you ask it for the Volume, you might get something else back if someone is controlling the receiver at the same time. Handling these cases would make the library much more complex for basically no added gain. (Sometimes keeping things simple is better than correctness.)
Also sending messages without waiting makes the receiver ignore them, etc...

I've experimented with a few ways of writing asyncio API in my gitlab repo [nadtcp](https://gitlab.com/mindig.marton/nadtcp) to increase stability (handling socket hangs, etc) / decrease complexity. I'm using it with Home Assistant [ha-nadtcp](https://gitlab.com/mindig.marton/ha-nadtcp), so I wanted to keep the integration as simple as possible and I don't think mixing asyncio / blocking io in the same integration would work. So basically this would split the Home Assistant integration into two...

Though I guess I could open a PR for the C338, then slowly merge the protocol (transport) implementation with the D7050 while keeping the quirky command handling separate.

**Comment by gladhorn** [comment]
Hi @mindigmarton :) Does that mean you actually are able to test the TCP class and related? I was wondering about trying to bring everything more in line since we right now effectively have two libraries in one repo...
It's a bit scary to try doing things without being able to confirm whether it's working or not.
I think this issue is not really something we can act on though, so we should maybe rather work in pull requests.
Ideally we'd end up with common public api which has different backends, maybe we could even have some of the TCP stuff added to the CMDS dictionary.
Independent of how we continue, I do think it's best to try keeping things in one place and trying to abstract away the differences. Hopefully we reach a point where we can fix bugs one place and improve support for several devices.

**Comment by mindigmarton** [comment]
I'd be vary of providing a unified API for all of them, should focus on first getting the C338 code in this repo, cleaning up the existing code (converting to async, etc). The receivers may share a common command set, but they don't act the same way, unfortunately, and I can't test the D7050, only the C338: (

**Comment by mindigmarton** [comment]
I've sold my NAD amplifier, so I can't continue working on this. Feel free to use whatever code from the C338 repository.
